### ISSUE TITLE
Erro to connect Failedtoscrapesocket:="read tcp 127.0.0.1:49884

### ISSUE THREAD

**Issue by ricardorrs** [letsencrypt/unbound_exporter#54]
Hey, guys,

I would like some help, I'm trying to configure the unbound_exporter but without success, some errors appeared, I'm reviewing the settings, but hours have passed and I'm tired, I have no light at the end of the tunnel

the alarm i'm getting is

May 30th 23:12:28 evrd-unbound unbound_exporter[788]: level=error ts=2023-05-31T03:12:28.226Z caller=unbound_exporter.go:461 Failedtoscrapesocket:="read tcp 127.0.0.1:39402- > 127.0.0.1:8953: read: connection reset by peer"
May 30th 23:12:38 evrd-unbound unbound_exporter[788]: level=error ts=2023-05-31T03:12:38.227Z caller=unbound_exporter.go:461 Failedtoscrapesocket:="read tcp 127.0.0.1:49884- > 127.0.0.1:8953: read: connection reset by peer"

Metric unbound_exporter

[METRIC.txt](https://github.com/letsencrypt/unbound_exporter/files/11608291/METRIC.txt)

Configuration unbound

[Config-unbound.txt](https://github.com/letsencrypt/unbound_exporter/files/11608279/Config-unbound.txt)

This is my .service that I created after compiling unbound_exporter.go

[prometheus-unbound-exporter.service.txt](https://github.com/letsencrypt/unbound_exporter/files/11608318/prometheus-unbound-exporter.service.txt)

_**I left it as simple as possible after all
They are default config**_
```
    -unbound.cert "/etc/unbound/unbound_control.pem"
     -unbound.key "/etc/unbound/unbound_control.key"
     -unbound.host "tcp://localhost:8953"
```

I didn't do many configurations in the DNS after all I'm testing the unbound_exporter, so far it hasn't been successful.
Inside the metric file the unbound_up 0 and this is not true but I believe that this parameter needs to be updated as soon as the unbound_exporter can connect to the 8953, just one more observation both the unbound and the exporter both are on the same linux "rockylinux9"

**Comment by pgporada** [comment]
TL;DR: You need a bit more configuration in your provided `prometheus-unbound-exporter.service` file. See [this issue.](https://github.com/letsencrypt/unbound_exporter/issues/44)

----

That's frustrating, been there before. We need to do some more digging before you get sorted out. 

First thing, I noticed that your systemd service file for the unbound exporter is lacking necessary config to talk to unbound. From your unbound config (comments and newlines removed for brevity):
```
remote-control:
        control-enable: yes
        control-interface: 0.0.0.0
        control-port: 8953
        control-use-cert: "no"         <========= This is the important bit =======
        server-key-file: "/etc/unbound/unbound_server.key"
        server-cert-file: "/etc/unbound/unbound_server.pem"
        control-key-file: "/etc/unbound/unbound_control.key"
        control-cert-file: "/etc/unbound/unbound_control.pem"
```

Using unbound-control which is a C utility packaged with unbound, can you get metrics? This is the mechanism that prometheus-unbound-exporter will use too. Using your config I get this output which is great. If you're getting errors, post them here.
```
$ unbound-control stats_noreset
thread0.num.queries=0
thread0.num.queries_ip_ratelimited=0
thread0.num.cachehits=0
...
```

Next, let's make sure that the unbound exporter is actually running. I built mine with go1.20.4. What go version did you build yours with?
```
$ systemctl status prometheus-unbound-exporter
● prometheus-unbound-exporter.service - Prometheus exporter for Unbound metrics, written in Go with pluggable metric collectors. The metrics exporter converts Unbound metric names to Prometheus metric names and labels by using a set of regular expressions.
     Loaded: loaded (/etc/systemd/system/prometheus-unbound-exporter.service; disabled; vendor preset: enabled)
     Active: active (running) since Wed 2023-05-31 12:04:49 EDT; 4s ago
       Docs: https://github.com/letsencrypt/unbound_exporter
   Main PID: 27661 (unbound_exporte)
      Tasks: 7 (limit: 76734)
     Memory: 1.9M
        CPU: 6ms
     CGroup: /system.slice/prometheus-unbound-exporter.service
             └─27661 /home/phil/.gvm/pkgsets/go1.20.4/global/bin/unbound_exporter

May 31 12:04:49 nerdwagon systemd[1]: Started Prometheus exporter for Unbound metrics, written in Go with pluggable metric collectors. The metrics exporter converts Unbound metric names to Prometheus metric names and labels by using a set of regular expressions..
May 31 12:04:49 nerdwagon unbound_exporter[27661]: level=info ts=2023-05-31T16:04:49.516Z caller=unbound_exporter.go:480 Startingunbound_exporter=(MISSING)
May 31 12:04:49 nerdwagon unbound_exporter[27661]: level=info ts=2023-05-31T16:04:49.517Z caller=unbound_exporter.go:498 Listeningonaddress:port>=:9167

```

We'll hit the prometheus-unbound-exporter metrics endpoint to see if we're getting anything other than just go stats. We won't,  my output is the same as yours, but here's the output for posterity.
```
$ curl -s localhost:9167/metrics | grep ^unbound
unbound_up 0
```

The telling output is 
```
caller=unbound_exporter.go:461 Failedtoscrapesocket:="read tcp 127.0.0.1:39402- > 127.0.0.1:8953: read: connection reset by peer"
```

What that means is you'll need a bit more configuration in your provided prometheus-unbound-exporter.service unit file. Specifically what this does is prevent prometheus-unbound-exporter from communicating over TLS to `unbound-control`. If you want to have TLS communication, I suggest using the files I've created in the [contrib](https://github.com/letsencrypt/unbound_exporter/tree/main/contrib) folder.
```
$ sudo vim /etc/systemd/system/prometheus-unbound-exporter.service

[Unit]
Description=Prometheus exporter for Unbound metrics, written in Go with pluggable metric collectors. The metrics exporter converts Unbound metric names to Prometheus metric names and labels by using a set of regular expressions.
Documentation=https://github.com/letsencrypt/unbound_exporter
After=network.target

[Service]
Type=simple
ExecStart=/usr/sbin/unbound_exporter -unbound.ca "" -unbound.cert ""

[Install]
WantedBy=multi-user.target
```

Then do some systemd shenanigans to restart and check the process.
```
sudo systemctl daemon-reload
sudo systemctl restart prometheus-unbound-exporter
systemctl status prometheus-unbound-exporter
```

When prometheus-unbound-exporter is wired up correctly, it'll output metrics as follows:
```
curl -s localhost:9167/metrics | grep ^unbound
unbound_answer_rcodes_total{rcode="FORMERR"} 0
unbound_answer_rcodes_total{rcode="NOERROR"} 0
unbound_answer_rcodes_total{rcode="NOTIMPL"} 0
unbound_answer_rcodes_total{rcode="NXDOMAIN"} 0
unbound_answer_rcodes_total{rcode="REFUSED"} 0
unbound_answer_rcodes_total{rcode="SERVFAIL"} 0
unbound_answers_bogus 0
...
```

**Comment by skeetmtp** [comment]
Thanks you saved my day! :)

To recap:

Mandatory lines in /etc/unbound/unbound.conf
```
remote-control:
  control-enable: yes
  control-use-cert: no
```

command line args:
```
unbound_exporter -unbound.ca "" -unbound.cert ""
```

With just this changes it works like a charm.
